cmd/sort/main.go

```diff
@@ -15,7 +17,20 @@ func main() {
 	//naive := parse.SortAlgorithm{FnAppend: tools.AppendWrapper, FnSort: sort.Lines}
 	binary := parse.SortAlgorithm{FnAppend: sort.BinarySearchAppend, FnSort: tools.EmptySortWrapper}
 
-	lines, err := parse.ScanLines(arg, binary)
+	files, err := arg.OpenFiles()
+	defer func() {
+		for _, f := range files {
+			if err := f.Close(); err != nil {
+				fmt.Println(err)
+				os.Exit(2)
+			}
+		}
+	}()
```

Comment:
Can't this closing of the files be at the end of the `arg.OpenFiles()` function? Now you always when using the function, have to remember that you have to close the files, even though the return value is `[]io.ReadCloser`. Just thoughts...

Reply:
If you mean something like:
```golang
func (o *Options) OpenFiles() ([]io.ReadCloser, error) {
	...
	defer func() {
		for _, f := range files {
			f.Close()
		}
	}()
	...
	return files, err
}
```
you would close `files` before using them, they will be no longer valid for the caller then.
We could just establish a contract between the callee and the caller: `OpenFiles()` will eventually return a list of valid `os.ReadCloser` or an error but it's the caller responsibility to handle their lifecycle.
`os.Open()` does something similar if you think about it.